src/vec_int512_runtime.c

```diff
@@ -0,0 +1,238 @@
+/*
+ Copyright (c) [2019] Steven Munroe.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+ vec_int512_runtime.c
+
+ Contributors:
+      Steven Munroe
+      Created on: Aug 20, 2019
+ */
+
+#include <pveclib/vec_int512_ppc.h>
+
+__VEC_U_256
+__VEC_PWR_IMP (vec_mul128x128) (vui128_t m1l, vui128_t m2l)
+{
+  return vec_mul128x128_inline (m1l, m2l);
+}
+
+__VEC_U_512
+__VEC_PWR_IMP (vec_mul256x256) (__VEC_U_256 m1, __VEC_U_256 m2)
+{
+  return vec_mul256x256_inline (m1, m2);
+}
+
+__VEC_U_640
+__VEC_PWR_IMP (vec_mul512x128) (__VEC_U_512 m1, vui128_t m2)
+{
+  return vec_mul512x128_inline (m1, m2);
+}
+
+__VEC_U_1024 __attribute__((flatten))
+__VEC_PWR_IMP (vec_mul512x512) (__VEC_U_512 m1, __VEC_U_512 m2)
+{
+  return vec_mul512x512_inline (m1, m2);
+}
+
+/** \brief Macros that invert the indexes so that quadword array endian
+ * follows quadword endian.
+ *
+ * The low order quadword index is always 0.
+ * The high order quadword index is 1, 3, 7.
+ *
+ */
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define __NDX2(__index) (__index)
+#define __NDX4(__index) (__index)
+#define __NDX8(__index) (__index)
+#else
+#define __NDX2(__index) (1 - __index)
+#define __NDX4(__index) (3 - __index)
+#define __NDX8(__index) (7 - __index)
+#endif
+
+void __attribute__((flatten ))
+#if 0
+/* The Compiler team strongly suggested using  restrict here
+ * but in practice (at least for GCC 8/9) we are better off without.
+ * Waiting for a bug fix.
+ */
+__VEC_PWR_IMP (vec_mul1024x1024) (__VEC_U_2048* restrict r2048,
+				  __VEC_U_1024* restrict m1_1024,
+				  __VEC_U_1024* restrict m2_1024)
+#else
+__VEC_PWR_IMP (vec_mul1024x1024) (__VEC_U_2048* r2048,
+				  __VEC_U_1024* m1_1024,
+				  __VEC_U_1024* m2_1024)
+#endif
+{
+  __VEC_U_512 *p2048, *m1, *m2;
+  __VEC_U_1024x512 subp0, subp1, subp2, subp3;
+  __VEC_U_512x1 sum0, sum1, sum2, sum3, sumx;
+  __VEC_U_512 temp[3];
+
+  p2048 = (__VEC_U_512 *) r2048;
+  m1 = (__VEC_U_512 *) m1_1024;
+  m2 = (__VEC_U_512 *) m2_1024;
+
+  subp0.x1024 = vec_mul512x512_inline (m1[__NDX2(0)], m2[__NDX2(0)]);
+  p2048[__NDX4(0)] = subp0.x2.v0x512;
+
+  subp1.x1024 = vec_mul512x512_inline (m1[__NDX2(1)], m2[__NDX2(0)]);
+  sum1.x640 = vec_add512cu (subp1.x2.v0x512, subp0.x2.v1x512);
+
+  temp[0] = sum1.x2.v0x512;
+  temp[1] = vec_add512ze (subp1.x2.v1x512, sum1.x2.v1x128);
+  COMPILE_FENCE;
```

Comment:
Do you want to mention here why the FENCE is needed?  This is to avoid spill, correct? Should the compiler handle this better, obviating the need?
_same comment for other uses of the fence herein_

Reply:
COMPILE_FENCE is documented in doxygen docs for vec_int512_ppc.h.

/*! \brief A compiler fence to prevent excessive code motion.
 *
 *  We use the COMPILER_FENCE to limit instruction scheduling
 *  and code motion to smaller code blocks. This in turn reduces
 *  register pressure and avoids generating spill code.
 */